InstructorCourseJoinAuthenticatedActionr: No institute found for the course

No institute found for the course


Actual user abc
Request Method GET
User Agent Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/601.2.7 (KHTML, like Gecko) Version/9.0.1 Safari/537.86.2
Request Url https://teammatesv4.appspot.com/page/instructorCourseJoinAuthenticated
Request Path /page
Request Parameters {key::abc3CF7B189E1C603F59BD79FC7FC9D076B6C55940983641A3B18FA8825D6D34FF1A8B755D36D1A5B2A2F8EC79B03872C877BCDF8C780ACE5F9A2AB0FCC8EBD9}
Stack Trace



java.lang.AssertionError: No institute found for the course
	at teammates.common.util.Assumption.fail(Assumption.java:59)
	at teammates.common.util.Assumption.assertTrue(Assumption.java:27)
	at teammates.common.util.Assumption.assertFalse(Assumption.java:44)
	at teammates.common.util.Assumption.assertNotEmpty(Assumption.java:370)
	at teammates.logic.core.AccountsLogic.getCourseInstitute(AccountsLogic.java:100)
	at teammates.logic.core.AccountsLogic.joinCourseForInstructorWithInstitute(AccountsLogic.java:181)
	at teammates.logic.core.AccountsLogic.joinCourseForInstructor(AccountsLogic.java:162)
	at teammates.logic.api.Logic.joinCourseForInstructor(Logic.java:581)
	at teammates.ui.controller.InstructorCourseJoinAuthenticatedAction.execute(InstructorCourseJoinAuthenticatedAction.java:35)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:337)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at teammates.ui.controller.ControllerServlet.doGet(ControllerServlet.java:38)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:143)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:468)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:256)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)

--- a/src/main/java/teammates/ui/controller/InstructorCourseInstructorDeleteAction.java
+++ b/src/main/java/teammates/ui/controller/InstructorCourseInstructorDeleteAction.java
@@ -27,20 +27,20 @@ public class InstructorCourseInstructorDeleteAction extends Action {
         new GateKeeper().verifyAccessible(
                 instructor, logic.getCourse(courseId), Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR);
 
+        // Check if there is a joined instructor (other than the instructor to delete) with the privilege of modifying instructors
         List<InstructorAttributes> instructors = logic.getInstructorsForCourse(courseId);
-        int numOfInstrCanModifyInstructor = 0;
-        InstructorAttributes instrCanModifyInstructor = null;
+        boolean hasInstrCanModifyInstructor = false;
         for (InstructorAttributes instr : instructors) {
-            if (instructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR)) {
-                numOfInstrCanModifyInstructor++;
-                instrCanModifyInstructor = instr;
+            if (instr.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR)
+                && !instr.getEmail().equals(instructorEmail)
+                && instr.isRegistered()) {
+                hasInstrCanModifyInstructor = true;
+                break;
             }
         }
-        boolean lastCanModifyInstructor = (numOfInstrCanModifyInstructor <= 1) && 
-                (instrCanModifyInstructor != null && instrCanModifyInstructor.email.equals(instructorEmail));
         
         /* Process deleting an instructor and setup status to be shown to user and admin */
-        if (!lastCanModifyInstructor) {
+        if (hasInstrCanModifyInstructor) {
             logic.deleteInstructor(courseId, instructorEmail);
             
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_INSTRUCTOR_DELETED, StatusMessageColor.SUCCESS));

--- a/src/main/java/teammates/ui/controller/InstructorCourseInstructorDeleteAction.java
+++ b/src/main/java/teammates/ui/controller/InstructorCourseInstructorDeleteAction.java
@@ -27,20 +27,8 @@ public class InstructorCourseInstructorDeleteAction extends Action {
         new GateKeeper().verifyAccessible(
                 instructor, logic.getCourse(courseId), Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR);
 
-        // Check if there is a joined instructor (other than the instructor to delete) with the privilege of modifying instructors
-        List<InstructorAttributes> instructors = logic.getInstructorsForCourse(courseId);
-        boolean hasInstrCanModifyInstructor = false;
-        for (InstructorAttributes instr : instructors) {
-            if (instr.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR)
-                && !instr.getEmail().equals(instructorEmail)
-                && instr.isRegistered()) {
-                hasInstrCanModifyInstructor = true;
-                break;
-            }
-        }
-        
         /* Process deleting an instructor and setup status to be shown to user and admin */
-        if (hasInstrCanModifyInstructor) {
+        if (hasAlternativeInstructor(courseId, instructorEmail)) {
             logic.deleteInstructor(courseId, instructorEmail);
             
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_INSTRUCTOR_DELETED, StatusMessageColor.SUCCESS));
@@ -66,4 +54,26 @@ public class InstructorCourseInstructorDeleteAction extends Action {
         return result;
     }
 
+    /**
+     * @param courseId is the id of the course
+     * @param instructorToDeleteEmail is the email of the instructor who is being deleted
+     * @return true if there is a joined instructor (other than the instructor to delete) with the privilege of modifying instructors
+     */
+    private boolean hasAlternativeInstructor(String courseId, String instructorToDeleteEmail) {
+
+        List<InstructorAttributes> instructors = logic.getInstructorsForCourse(courseId);
+
+        for (InstructorAttributes instr : instructors) {
+
+            boolean isAlternativeInstructor = instr.isRegistered()
+                                              && !instr.getEmail().equals(instructorToDeleteEmail)
+                                              && instr.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR);
+
+            if (isAlternativeInstructor) {
+                return true;
+            }
+        }
+
+        return false;
+    }
 }

--- a/src/test/java/teammates/test/cases/ui/browsertests/InstructorCourseEditPageUiTest.java
+++ b/src/test/java/teammates/test/cases/ui/browsertests/InstructorCourseEditPageUiTest.java
@@ -11,6 +11,8 @@ import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
 import teammates.common.datatransfer.DataBundle;
+import teammates.common.datatransfer.InstructorAttributes;
+import teammates.common.datatransfer.InstructorPrivileges;
 import teammates.common.util.AppUrl;
 import teammates.common.util.Const;
 import teammates.common.util.FieldValidator;
@@ -418,37 +420,61 @@ public class InstructorCourseEditPageUiTest extends BaseUiTestCase {
         
         ______TS("delete instructor then cancel");
         
-        courseEditPage.clickDeleteInstructorLinkAndCancel();
+        courseEditPage.clickDeleteInstructorLinkAndCancel(1);
         assertNotNull(BackDoor.getInstructorAsJsonByGoogleId(instructorId, courseId));
         
         ______TS("delete instructor successfully");
         
-        courseEditPage.clickDeleteInstructorLinkAndConfirm();
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(1);
         String expectedMsg = "The instructor has been deleted from the course.";
         courseEditPage.verifyStatus(expectedMsg);
         
-        ______TS("delete all instructors");
+        ______TS("delete all other instructors");
         
-        courseEditPage.clickDeleteInstructorLinkAndConfirm();
-        courseEditPage.clickDeleteInstructorLinkAndConfirm();
-        courseEditPage.clickDeleteInstructorLinkAndConfirm();
-        courseEditPage.clickDeleteInstructorLinkAndConfirm();
-        courseEditPage.clickDeleteInstructorLinkAndConfirm();
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(1);
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(1);
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(1);
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(1);
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(1);
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(2);
         
-        ______TS("deleted own instructor role and redirect to courses page");
+        ______TS("test the only registered instructor with the privilege to modify instructors cannot be deleted");
         
-        // Change login id to another instructor
+        // Create an registered instructor with all privileges except modifying instructors
+        InstructorPrivileges privilege = new InstructorPrivileges(Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_COOWNER);
+        privilege.updatePrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_INSTRUCTOR, false);
+        InstructorAttributes instructor = new InstructorAttributes("InsCrsEdit.reg", courseId, "Teammates Reg", "InsCrsEdit.reg@gmail.tmt", 
+                                        Const.InstructorPermissionRoleNames.INSTRUCTOR_PERMISSION_ROLE_CUSTOM, "Teammates Reg", privilege);
+        BackDoor.createInstructor(instructor);
+        
+        // Create an unregistered instructor with co-owner privilege
+        courseEditPage.addNewInstructor("Unreg Instructor", "InstructorCourseEditEmail@gmail.tmt");
+        
+        // Delete own instructor role
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(2);
+        assertEquals(Const.StatusMessages.COURSE_INSTRUCTOR_DELETE_NOT_ALLOWED, courseEditPage.getStatus());
+        
+        // Delete other instructors
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(3);
+        assertEquals(Const.StatusMessages.COURSE_INSTRUCTOR_DELETED, courseEditPage.getStatus());
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(1);
+        assertEquals(Const.StatusMessages.COURSE_INSTRUCTOR_DELETED, courseEditPage.getStatus());
+        
+        ______TS("delete own instructor role and redirect to courses page");
+        
+        // Create another registered instructor with co-owner privilege
         BackDoor.createInstructor(testData.instructors.get("InsCrsEdit.coord"));
-        instructorId = testData.instructors.get("InsCrsEdit.coord").googleId;
-        courseEditPage.clickDeleteInstructorLinkAndConfirm();
+        courseEditPage = getCourseEditPage();
+        
+        // Delete own instructor role
+        courseEditPage.clickDeleteInstructorLinkAndConfirm(2);
 
         InstructorCoursesPage coursesPage = courseEditPage.changePageType(InstructorCoursesPage.class);
         coursesPage.waitForAjaxLoadCoursesSuccess();
         coursesPage.verifyStatus(Const.StatusMessages.COURSE_INSTRUCTOR_DELETED + "\n"
                                  + Const.StatusMessages.COURSE_EMPTY);
         
-        // Change back login id to original instructor to ensure remaining test cases work properly
-        instructorId = testData.instructors.get("InsCrsEdit.test").googleId;
+        // Restore own instructor role to ensure remaining test cases work properly
         BackDoor.createInstructor(testData.instructors.get("InsCrsEdit.test"));
     }
     
@@ -473,7 +499,7 @@ public class InstructorCourseEditPageUiTest extends BaseUiTestCase {
         ______TS("make a new unregistered instructor and test that its email can't be edited");
         
         courseEditPage.addNewInstructor("Unreg Instructor", "InstructorCourseEditEmail@gmail.tmt");
-        int unregInstrNum = 4;
+        int unregInstrNum = 3;
         assertEquals("Unreg Instructor", courseEditPage.getNameField(unregInstrNum).getAttribute("value"));
         assertFalse(courseEditPage.getNameField(unregInstrNum).isEnabled());
         

--- a/src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java
+++ b/src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java
@@ -21,9 +21,6 @@ public class InstructorCourseEditPage extends AppPage {
     @FindBy(id = "instrEditLink1")
     private WebElement editInstructorLink;
     
-    @FindBy(id = "instrDeleteLink1")
-    private WebElement deleteInstructorLink;
-    
     @FindBy(id = "instrRemindLink4")
     private WebElement inviteInstructorLink;
 
@@ -315,12 +312,22 @@ public class InstructorCourseEditPage extends AppPage {
         clickAndCancel(deleteCourseLink);
     }
     
-    public void clickDeleteInstructorLinkAndConfirm() {
+    /**
+     * Click the delete instructor button at position {@code instrNum} and click "Yes" in the follow up dialog
+     * @param instrNum is the position of the instructor (e.g. 1, 2, 3, ...)
+     */
+    public void clickDeleteInstructorLinkAndConfirm(int instrNum) { 
+        WebElement deleteInstructorLink = browser.driver.findElement(By.id("instrDeleteLink" + instrNum));
         clickAndConfirm(deleteInstructorLink);
         waitForPageToLoad();
     }
     
-    public void clickDeleteInstructorLinkAndCancel() {
+    /**
+     * Click the delete instructor button at position {@code instrNum} and click "No" in the follow up dialog
+     * @param instrNum is the position of the instructor (e.g. 1, 2, 3, ...)
+     */
+    public void clickDeleteInstructorLinkAndCancel(int instrNum) {
+        WebElement deleteInstructorLink = browser.driver.findElement(By.id("instrDeleteLink" + instrNum));
         clickAndCancel(deleteInstructorLink);
     }
     

